src/Microsoft.SqlTools.ServiceLayer/Scripting/ScriptingObjectMatchProcessor.cs

```diff
@@ -0,0 +1,111 @@
+            IEnumerable<ScriptingObject> candidates)
+        {
+            if (candidates == null)
+            {
+                throw new ArgumentNullException("candidates");
+            }
+
+            IEnumerable<ScriptingObject> matchedObjects = new List<ScriptingObject>();
+
+            if (includeCriteria != null && includeCriteria.Count() > 0)
+            {
+                foreach (ScriptingObject scriptingObjectCriteria in includeCriteria)
+                {
+                    IEnumerable<ScriptingObject> matches = MatchCriteria(scriptingObjectCriteria, candidates);
+                    matchedObjects = matchedObjects.Union(matches);
+                }
+            }
+            else
+            {
+                matchedObjects = candidates;
+            }
+
+            if (excludeCriteria != null)
+            {
+                foreach (ScriptingObject scriptingObjectCriteria in excludeCriteria)
+                {
+                    IEnumerable<ScriptingObject> matches = MatchCriteria(scriptingObjectCriteria, candidates);
+                    matchedObjects = matchedObjects.Except(matches);
+                }
+            }
+
+            return matchedObjects;
+        }
+
+        private static IEnumerable<ScriptingObject> MatchCriteria(ScriptingObject criteria, IEnumerable<ScriptingObject> databaseObjects)
+        {
+            IEnumerable<ScriptingObject> matchedObjects = databaseObjects;
+
+            if (!string.IsNullOrWhiteSpace(criteria.Type))
+            {
+                matchedObjects = matchedObjects.Where(o => string.Equals(criteria.Type, o.Type, StringComparison.OrdinalIgnoreCase));
+            }
+
+            if (!string.IsNullOrWhiteSpace(criteria.Schema))
+            {
+                if (criteria.Schema.Equals("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    // Don't filter any objects
+                }
+                if (criteria.Schema.EndsWith("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    matchedObjects = matchedObjects.Where(o => o.Schema.StartsWith(
+                        criteria.Schema.Substring(0, criteria.Schema.Length - 1), 
+                        StringComparison.CurrentCultureIgnoreCase));
+                }
+                else
+                {
+                    matchedObjects = matchedObjects.Where(o => string.Equals(criteria.Schema, o.Schema, StringComparison.OrdinalIgnoreCase));
+                }
+            }
+
+            if (!string.IsNullOrWhiteSpace(criteria.Name))
+            {
+                if (criteria.Name.Equals("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    // Don't filter any objects
+                }
+                else if (criteria.Name.EndsWith("*", StringComparison.OrdinalIgnoreCase))
+                {
+                    matchedObjects = matchedObjects.Where(o => o.Name.StartsWith(
```

Comment:
Yep, good idea.

---
In reply to: [111654162](https://github.com/Microsoft/sqltoolsservice/pull/316#discussion_r111654162) [](ancestors = 111654162)